contracts/transfer/circuits/src/gadgets/schnorr.rs

```diff
@@ -0,0 +1,206 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_plonk::constraint_system::ecc::scalar_mul::fixed_base::scalar_mul as fixed_base_scalar_mul;
+use dusk_plonk::constraint_system::ecc::scalar_mul::variable_base::variable_base_scalar_mul;
+use dusk_plonk::constraint_system::ecc::Point as PlonkPoint;
+use dusk_plonk::jubjub::{
+    AffinePoint, ExtendedPoint, GENERATOR_EXTENDED, GENERATOR_NUMS_EXTENDED,
+};
+use dusk_plonk::prelude::*;
+use plonk_gadgets::AllocatedScalar;
+use poseidon252::sponge::sponge::sponge_hash_gadget;
+use std::convert::TryInto;
+
+/// Utilises a Schnorr signature scheme,
+/// to prove the knowledge of a discrete
+/// log for a given public key.
+#[allow(non_snake_case)]
+pub fn schnorr_one_key(
+    composer: &mut StandardComposer,
+    signature: AllocatedScalar,
+    R: PlonkPoint,
+    pk: PlonkPoint,
+    message: AllocatedScalar,
+) {
+    let h = sponge_hash_gadget(composer, &[message.var]);
+    let c = sponge_hash_gadget(composer, &[*R.x(), *R.y(), h]);
+    let b = BlsScalar::zero();
+    let b = composer.add_witness_to_circuit_description(b);
+
+    let challenge = composer.xor_gate(c, b, 250);
+
+    let sig =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let p = variable_base_scalar_mul(composer, challenge, pk);
+
+    let add = sig.point().add(composer, *p.point());
+    composer.assert_equal_point(add, R);
+}
+
+/// Utilises a Schnorr signature scheme,
+/// to prove the knowledge of the discrete
+/// log for given keys in a public key pair.
+/// Also verifying that both keys share the
+/// same discrete log.
+#[allow(non_snake_case)]
+pub fn schnorr_two_keys(
+    composer: &mut StandardComposer,
+    signature: AllocatedScalar,
+    R: PlonkPoint,
+    R_prime: PlonkPoint,
+    pk: PlonkPoint,
+    pk_prime: PlonkPoint,
+    message: AllocatedScalar,
+) {
+    let h = sponge_hash_gadget(composer, &[message.var]);
+    let c = sponge_hash_gadget(
+        composer,
+        &[*R.x(), *R.y(), *R_prime.x(), *R_prime.y(), h],
+    );
+    let b = BlsScalar::zero();
+    let b = composer.add_witness_to_circuit_description(b);
+
+    let challenge = composer.xor_gate(c, b, 250);
+    let sig_1 =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let sig_2 =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_NUMS_EXTENDED);
+    let pub_1 = variable_base_scalar_mul(composer, challenge, pk);
+    let pub_2 = variable_base_scalar_mul(composer, challenge, pk_prime);
+
+    let add_1 = sig_1.point().add(composer, *pub_1.point());
+    composer.assert_equal_point(add_1, R);
+    let add_2 = sig_2.point().add(composer, *pub_2.point());
+    composer.assert_equal_point(add_2, R_prime);
```

Comment:
Same as above. If the points are public, then we need to add the checks outside.

Reply:
Nothing is public according to Tog.